opa-audit-operator/src/charm.py

```diff
@@ -0,0 +1,177 @@
+                    "files/constrainttemplates.templates.gatekeeper.sh.yaml",
+                    "files/constraintpodstatuses.status.gatekeeper.sh.yaml",
+                    "files/constrainttemplatepodstatuses.status.gatekeeper.sh.yaml",
+                    # "files/sync.yaml",
+                ]
+            ]
+        except yaml.YAMLError as exc:
+            print("Error in configuration file:", exc)
+
+        crd_objects = [
+            CustomResourceDefintion(crd["metadata"]["name"], crd["spec"])
+            for crd in crds
+        ]
+
+        config = self.model.config
+        spec_template = {}
+        with open("files/pod-spec.yaml.jinja2") as fh:
+            spec_template = Template(fh.read())
+
+        import pdb; pdb.set_trace()
+        template_args = {
+            "crds": crd_objects,
+            "image_path": config["imagePath"],
+            "imagePullPolicy": config["imagePullPolicy"],
+            "app_name": self.app.name,
+            "audit_cli_args": self._audit_cli_args(),
+            "namespace": os.environ["JUJU_MODEL_NAME"],
+        }
+
+        spec = yaml.load(spec_template.render(**template_args))
+
+        print(f"Pod spec: {spec}")
+        return spec
+
+    def _audit_cli_args(self):
+        """
+        Construct command line arguments for OPA Audit
+        """
+
+        args = [
+            "--operation=audit",
+            "--operation=status",
+            "--logtostderr",
+        ]
+
+        return args
+
+    def _opa_config(self):
+        """
+        Construct opa configuration
+        """
+        config = self.model.config
+
+        logger.debug("opa config : {}".format(config))
+
+        return yaml.dump(config)
+
+    def _check_config(self):
+        """
+        Identify missing but required items in configuation
+        :returns: list of missing configuration items (configuration keys)
+        """
+        logger.debug("Checking Config")
+        config = self.model.config
+        missing = []
+
+        if not config.get("imagePath"):
+            missing.append("imagePath")
+
+        return missing
+
+    def _configure_pod(self):
+        """
+        Setup a new opa pod specification
+        """
+        logger.debug("Configuring Pod")
+        missing_config = self._check_config()
+        if missing_config:
+            logger.error(
+                "Incomplete Configuration : {}. "
+                "Application will be blocked.".format(missing_config)
+            )
+            self.unit.status = BlockedStatus(
+                "Missing configuration: {}".format(missing_config)
+            )
+            return
+
+        if not self.unit.is_leader():
+            self.unit.status = ActiveStatus()
+            return
+
+        self.unit.status = MaintenanceStatus("Setting pod spec.")
+        pod_spec = self._build_pod_spec()
+
+        self.model.pod.set_spec(pod_spec)
+        self.app.status = ActiveStatus()
```

Comment:
I don't think you should set app.status here. Just set the unit status and let Juju set the application status appropriately.

Reply:
Please remove this.

Reply:
removed